gadget/gadget.go

```diff
@@ -937,32 +937,45 @@ func IsCompatible(current, new *Info) error {
 	return nil
 }
 
-// LaidOutVolumeFromGadget takes a gadget rootdir and lays out the
-// partitions as specified.
-func LaidOutVolumeFromGadget(gadgetRoot string, model Model) (*LaidOutVolume, error) {
-	info, err := ReadInfo(gadgetRoot, model)
+// LaidOutSystemVolumeFromGadget takes a gadget rootdir and lays out the
+// partitions as specified. It returns one specific volume, which is the volume
+// on which system-* roles/partitions exist, all other volumes are assumed to
+// already be flashed and managed separately at image build/flash time, while
+// the ubuntu-* roles can be manipulated on the returned volume during install
+// mode.
+func LaidOutSystemVolumeFromGadget(gadgetRoot string, model Model) (*LaidOutVolume, error) {
+	// rely on the basic validation from ReadInfo to ensure that the ubuntu-*
+	// roles are all on the same volume for example
+	info, err := ReadInfoAndValidate(gadgetRoot, model, nil)
 	if err != nil {
 		return nil, err
 	}
-	// Limit ourselves to just one volume for now.
-	if len(info.Volumes) != 1 {
-		return nil, fmt.Errorf("cannot position multiple volumes yet")
-	}
 
 	constraints := LayoutConstraints{
 		NonMBRStartOffset: 1 * quantity.OffsetMiB,
-		SectorSize:        512,
+		// TODO:UC20: make SectorSize dynamic, either through config in the
+		//            gadget or by dynamic detection
+		SectorSize: sectorSize,
 	}
 
+	// find the volume with the ubuntu-boot role on it, we already validated
+	// that the ubuntu-* roles are all on the same volume
 	for _, vol := range info.Volumes {
-		pvol, err := LayoutVolume(gadgetRoot, vol, constraints)
-		if err != nil {
-			return nil, err
+		for _, structure := range vol.Structure {
+			// use the system-boot role
+			if structure.Role == SystemBoot || structure.Label == SystemBoot {
```

Comment:
we don't need the || Label part, because code in read will assign the Role implicitly if that label is used (for UC16/18)

Reply:
mmm I think you're right, let me double-check this, I had some issue with just using the Role, but that may have been because I first tried using system-data which isn't always in the gadget because it's an implicit structure for uc1{6,8}

Reply:
I don't think that's how it works in practice actually, I tried removing the `Label == SystemBoot` part and it fails in unit tests:
```
----------------------------------------------------------------------
FAIL: gadget_test.go:2032: gadgetYamlTestSuite.TestLaidOutSystemVolumeFromGadgetHappy

gadget_test.go:2041:
    c.Assert(err, IsNil)
... value *errors.errorString = &errors.errorString{s:"internal error: gadget passed validation but does not have system-* roles on any volume"} ("internal error: gadget passed validation but does not have system-* roles on any volume")

OOPS: 244 passed, 1 FAILED
--- FAIL: TestRun (1.70s)
FAIL
FAIL    github.com/snapcore/snapd/gadget        1.702s
```

I'll have a look at what's missing here

Reply:
the Label part is removed and the failing test was updated to use a non-nil model